File: backend/app/schemas/callback.py
from datetime import datetime
from enum import Enum
from typing import Any

from pydantic import BaseModel, Field


class CallbackStatus(str, Enum):
    """Callback status enum."""

    ACCEPTED = "accepted"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"


class TodoItem(BaseModel):
    """Todo item."""

    content: str
    status: str
    active_form: str | None = None


class McpStatus(BaseModel):
    """MCP status."""

    server_name: str
    status: str
    message: str | None = None


class FileChange(BaseModel):
    """File change record."""

    path: str
    status: str  # "added" | "modified" | "staged" | "deleted" | "renamed"
    added_lines: int = 0
    deleted_lines: int = 0
    diff: str | None = None
    old_path: str | None = None


class WorkspaceState(BaseModel):
    """Workspace state."""

    repository: str | None = None
    branch: str | None = None
    total_added_lines: int = 0
    total_deleted_lines: int = 0
    file_changes: list[FileChange] = Field(default_factory=list)
    last_change: datetime


class AgentCurrentState(BaseModel):
    """Agent current state."""

    todos: list[TodoItem] = Field(default_factory=list)
    mcp_status: list[McpStatus] = Field(default_factory=list)
    workspace_state: WorkspaceState | None = None
    current_step: str | None = None


class AgentCallbackRequest(BaseModel):
    """Agent execution callback request."""

    session_id: str
    time: datetime
    status: CallbackStatus
    progress: int
    new_message: Any | None = None
    state_patch: AgentCurrentState | None = None
    sdk_session_id: str | None = None
    workspace_files_prefix: str | None = None
    workspace_manifest_key: str | None = None
    workspace_archive_key: str | None = None
    workspace_export_status: str | None = None


class CallbackResponse(BaseModel):
    """Callback response."""

    session_id: str
    status: str
    callback_status: CallbackStatus | None = None
    message: str | None = None

File: frontend/features/chat/types/api/callback.ts
/**
 * Callback-related API types matching backend schemas
 */

export type CallbackStatus = "accepted" | "running" | "completed" | "failed";

export interface TodoItem {
  content: string;
  status: string;
  active_form?: string | null;
}

export interface McpStatus {
  server_name: string;
  status: string;
  message?: string | null;
}

// Type alias for backward compatibility
export type McpStatusItem = McpStatus;

export type FileChangeStatus = "added" | "modified" | "deleted" | "renamed";

export interface FileChange {
  path: string;
  status: FileChangeStatus | string;
  added_lines?: number;
  deleted_lines?: number;
  diff?: string | null;
  old_path?: string | null;
}

export interface WorkspaceState {
  repository?: string | null;
  branch?: string | null;
  total_added_lines?: number;
  total_deleted_lines?: number;
  file_changes?: FileChange[];
  last_change: string; // ISO datetime
}

export interface AgentCurrentState {
  todos?: TodoItem[];
  mcp_status?: McpStatus[];
  workspace_state?: WorkspaceState | null;
  current_step?: string | null;
}

/**
 * State patch for incremental session updates
 * Contains agent execution state that can be merged into session state
 */
export interface ApiStatePatch {
  todos?: TodoItem[];
  mcp_status?: McpStatus[];
  workspace_state?: WorkspaceState | null;
  current_step?: string | null;
}

export interface AgentCallbackRequest {
  session_id: string;
  time: string; // ISO datetime
  status: CallbackStatus;
  progress: number;
  new_message?: unknown | null;
  state_patch?: ApiStatePatch | null;
  sdk_session_id?: string | null;
}

export interface CallbackResponse {
  session_id: string;
  status: string;
  callback_status?: CallbackStatus | null;
  message?: string | null;
}
